File: Python/79.Word Search.py
#在二维字母矩阵中寻找给定的字母串，两个字母相邻即可(上下左右)
#深度优先搜索即可
class Solution:
    def exist(self, board, word):
        self.h = len(board)
        self.w = len(board[0])
        self.covered = [[0 for i in range(self.w)] for j in range(self.h)]
        if len(word)==0:
            return False
        for i in range(self.h):
            for j in range(self.w):
                if board[i][j]==word[0]:
                    if self.dfs(i,j,0,board,word)==True:
                        return True
        return False

    def dfs(self,i,j,k,board,word):
        if k==len(word)-1:
            return True
        self.covered[i][j] = 1
        if i>0 and self.covered[i-1][j]==0 and board[i-1][j]==word[k+1]:
            if self.dfs(i-1,j,k+1,board,word)==True:
                return True
        if j>0 and self.covered[i][j-1]==0 and board[i][j-1]==word[k+1]:
            if self.dfs(i,j-1,k+1,board,word)==True:
                return True
        if i<self.h-1 and self.covered[i+1][j]==0 and board[i+1][j]==word[k+1]:
            if self.dfs(i+1,j,k+1,board,word)==True:
                return True
        if j<self.w-1 and self.covered[i][j+1]==0 and board[i][j+1]==word[k+1]:
            if self.dfs(i,j+1,k+1,board,word)==True:
                return True
        self.covered[i][j] = 0
        return False


s = Solution()
print(s.exist([['A','B','C','E'],['B','F','C','S'],['A','D','E','E']],"ABADEC"))

File: C++/Backtracking/79.Word Search.cpp
/*
在二维字母矩阵中寻找给定的字母串，两个字母相邻即可(上下左右)。
*/
/*
思路：DFS递归。每次把之前遍历过的位置改为' '字符，防止再次遍历，递归过后再改回来。
*/
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    bool exist(vector<vector<char>>& board, string word) {
        for(int i=0 ; i < board.size(); ++i)
        {
        	for(int j=0 ; j < board[i].size(); ++j)
        	{
        		if(board[i][j]==word[0])
        		{
        			board[i][j] = ' ';
        			if(helpExist(board,word,i,j,1)==true)	return true;
        			board[i][j] = word[0];
        		}	
        	}
        }
        return false;
    }
    bool helpExist(vector<vector<char>>& board,const string& word,int row,int col,int index)
    {
    	if(index==word.size())	return true;
    	if(row>0 && board[row-1][col]==word[index])
    	{
    		board[row-1][col] = ' ';
    		if(helpExist(board,word,row-1,col,index+1)==true)	return true;
    		board[row-1][col] = word[index];
    	}
    	if(col>0 && board[row][col-1]==word[index])
    	{
    		board[row][col-1] = ' ';
    		if(helpExist(board,word,row,col-1,index+1)==true)	return true;
    		board[row][col-1] = word[index];
    	}
    	if(row<board.size()-1 && board[row+1][col]==word[index])
    	{
    		board[row+1][col] = ' ';
    		if(helpExist(board,word,row+1,col,index+1)==true)	return true;
    		board[row+1][col] = word[index];
    	}
    	if(col<board[0].size()-1 && board[row][col+1]==word[index])
    	{
    		board[row][col+1] = ' ';
    		if(helpExist(board,word,row,col+1,index+1)==true)	return true;
    		board[row][col+1] = word[index];
    	}
    	return false;
    }
};
int main()
{
	Solution s;
	vector<vector<char>> board{{'A','B','C','E'},{'S','F','C','S'},{'A','D','E','E'}};
    cout << s.exist(board,"ABCB");
}

